File: tests/stress_complex_json_tests.rs
use lorealdb::db_engine::DBEngine;
use pyo3::prelude::*;
use std::sync::Arc;
use std::time::{Duration, Instant};
use tempfile::tempdir;

// ---------------------------------------------------------
// Payload Generators
// ---------------------------------------------------------

fn generate_deeply_nested_object(depth: usize) -> String {
    let mut json = String::from("\"end_value\"");
    for i in 0..depth {
        json = format!("{{\"level_{}\": {}}}", depth - i, json);
    }
    json
}

fn generate_deeply_nested_array(depth: usize) -> String {
    let mut json = String::from("\"inner_item\"");
    for _ in 0..depth {
        json = format!("[{}]", json);
    }
    format!("{{\"deep_array\": {}}}", json)
}

fn generate_wide_json(key_count: usize) -> String {
    let mut pairs = Vec::with_capacity(key_count);
    for i in 0..key_count {
        pairs.push(format!("\"key_{}\": {}", i, i));
    }
    format!("{{{}}}", pairs.join(", "))
}

fn generate_large_llm_response() -> String {
    let citations: Vec<String> = (0..50)
        .map(|i| {
            format!(
                "{{\"id\": \"ref_{}\", \"url\": \"https://example.com/page_{}\", \"confidence\": 0.99, \"snippets\": [\"Matched text A\", \"Matched text B\"]}}",
                i, i
            )
        })
        .collect();

    format!(
            "{{\n  \"model\": \"gpt-4-turbo\",\n  \"choices\": [{{\n    \"message\": {{\n      \"role\": \"assistant\",\n      \"content\": \"This is a very long generated response...\",\n      \"metadata\": {{\n        \"citations\": [{}],\n        \"tokens_used\": 4500,\n        \"finish_reason\": \"stop\"\n      }}\n    }}\n  }}]\n}}",
            citations.join(",\n        ")

    )
}

#[test]
fn benchmark_complex_workload_and_concurrency() {
    // 1. Initialize Python and DB
    Python::initialize();
    let dir = tempdir().expect("Failed to create temp dir");
    let db_path = dir.path().join("production_stress_test.redb");

    // Wrap the engine in an Arc so we can share it across reader/writer threads
    let engine = Arc::new(DBEngine::new(db_path.to_str().unwrap()).expect("Failed to init engine"));

    // 2. Pre-generate payloads to avoid timing string creation
    println!("Generating complex payloads...");
    let deep_obj = generate_deeply_nested_object(150); // 150 levels deep
    let deep_arr = generate_deeply_nested_array(150);
    let wide_obj = generate_wide_json(1000); // 1000 flat keys
    let llm_resp = generate_large_llm_response();

    let payloads = vec![deep_obj, deep_arr, wide_obj, llm_resp];
    let total_operations = 50_000;

    println!("Starting Concurrent Read/Write Stress Test...");
    let system_start = Instant::now();

    // 3. Spawn Reader Threads (Simulating concurrent API GET requests)
    let mut read_handles = vec![];
    for thread_id in 0..4 {
        let engine_clone = Arc::clone(&engine);

        let handle = std::thread::spawn(move || {
            // Each thread does 12,500 reads
            let reads_per_thread = total_operations / 4;

            Python::attach(|py| {
                for i in 0..reads_per_thread {
                    // Pseudo-randomly pick an ID that might or might not exist yet
                    // using deterministic math to avoid requiring the `rand` crate.
                    let target_id_num = (i * 997 + thread_id) % total_operations;
                    let target_id = format!("doc_{}", target_id_num);

                    // We don't unwrap the inner value, because the writer might not
                    // have inserted it yet. We just want to stress the read transaction lock.
                    let _ = engine_clone
                        .get(py, &target_id)
                        .expect("Read transaction panicked");
                }
            });
        });
        read_handles.push(handle);
    }

    // 4. Run Writer in the main thread (Simulating API POST requests)
    let write_start = Instant::now();
    for i in 0..total_operations {
        let id = format!("doc_{}", i);
        // Round-robin through the 4 complex payloads
        let payload = &payloads[i % 4];

        engine
            .insert_json(&id, payload)
            .expect("Queue failed to accept insert");
    }
    let queue_time = write_start.elapsed();

    // 5. Wait for readers to finish their spam
    for handle in read_handles {
        handle.join().expect("Reader thread panicked!");
    }

    // 6. Polling Loop: Wait for the background worker to flush the final write
    let final_id = format!("doc_{}", total_operations - 1);
    let commit_start = Instant::now();
    loop {
        let is_committed = Python::attach(|py| engine.get(py, &final_id).unwrap().is_some());

        if is_committed {
            break;
        }
        std::thread::sleep(Duration::from_millis(10));
    }

    let total_time = system_start.elapsed();

    println!("-----------------------------------------------");
    println!(
        "Total Operations:      {} Writes, {} Reads",
        total_operations, total_operations
    );
    println!("Main Thread Write Q:   {:?}", queue_time);
    println!("Background I/O Wait:   {:?}", commit_start.elapsed());
    println!("Total System Time:     {:?}", total_time);
    println!("-----------------------------------------------");
    println!(
        "Mixed Throughput:      {:.0} ops/sec",
        (total_operations * 2) as f64 / total_time.as_secs_f64()
    );
}

File: benchmarks/stress_complex_json_tests.py
import os
import tempfile
import time
from concurrent.futures import ThreadPoolExecutor

from lorealdb import DBEngine

# ---------------------------------------------------------
# Payload Generators
# ---------------------------------------------------------


def generate_deeply_nested_object(depth: int) -> str:
    json_str = '"end_value"'
    for i in range(depth):
        json_str = f'{{"level_{depth - i}": {json_str}}}'
    return json_str


def generate_deeply_nested_array(depth: int) -> str:
    json_str = '"inner_item"'
    for _ in range(depth):
        json_str = f"[{json_str}]"
    return f'{{"deep_array": {json_str}}}'


def generate_wide_json(key_count: int) -> str:
    pairs = [f'"key_{i}": {i}' for i in range(key_count)]
    return f"{{{', '.join(pairs)}}}"


def generate_large_llm_response() -> str:
    citations = []
    for i in range(50):
        citations.append(
            f'{{"id": "ref_{i}", "url": "https://example.com/page_{i}", '
            f'"confidence": 0.99, "snippets": ["Matched text A", "Matched text B"]}}'
        )

    citations_str = ",\n        ".join(citations)
    return f"""{{
  "model": "gpt-4-turbo",
  "choices": [{{
    "message": {{
      "role": "assistant",
      "content": "This is a very long generated response...",
      "metadata": {{
        "citations": [{citations_str}],
        "tokens_used": 4500,
        "finish_reason": "stop"
      }}
    }}
  }}]
}}"""


# ---------------------------------------------------------
# Reader Worker
# ---------------------------------------------------------


def reader_worker(
    engine: DBEngine, thread_id: int, num_reads: int, total_operations: int
):
    for i in range(num_reads):
        # Pseudo-randomly pick an ID that might or might not exist yet
        # using deterministic math to match the Rust implementation.
        target_id_num = (i * 997 + thread_id) % total_operations
        target_id = f"doc_{target_id_num}"

        # We just want to stress the read transaction lock.
        engine.get(target_id)


def benchmark_complex_workload_and_concurrency():
    # 1. Initialize DB
    with tempfile.TemporaryDirectory() as temp_dir:
        db_path = os.path.join(temp_dir, "production_stress_test.redb")
        engine = DBEngine(db_path)

        # 2. Pre-generate payloads to avoid timing string creation
        print("Generating complex payloads...")
        deep_obj = generate_deeply_nested_object(100)  # Changed from 150
        deep_arr = generate_deeply_nested_array(100)  # Changed from 150
        wide_obj = generate_wide_json(1000)  # 1000 flat keys
        llm_resp = generate_large_llm_response()

        payloads = [deep_obj, deep_arr, wide_obj, llm_resp]
        total_operations = 50_000

        print("Starting Concurrent Read/Write Stress Test...")
        system_start = time.perf_counter()

        # 3. Spawn Reader Threads (Simulating concurrent API GET requests)
        reads_per_thread = total_operations // 4
        executor = ThreadPoolExecutor(max_workers=4)
        futures = []
        for thread_id in range(4):
            futures.append(
                executor.submit(
                    reader_worker, engine, thread_id, reads_per_thread, total_operations
                )
            )

        # 4. Run Writer in the main thread (Simulating API POST requests)
        write_start = time.perf_counter()
        for i in range(total_operations):
            doc_id = f"doc_{i}"
            # Round-robin through the 4 complex payloads
            payload = payloads[i % 4]
            engine.insert_json(doc_id, payload)

        queue_time = time.perf_counter() - write_start

        # 5. Wait for readers to finish their spam
        for future in futures:
            future.result()
        executor.shutdown()

        # 6. Polling Loop: Wait for the background worker to flush the final write
        final_id = f"doc_{total_operations - 1}"
        commit_start = time.perf_counter()
        while True:
            if engine.get(final_id) is not None:
                break
            time.sleep(0.01)

        total_time = time.perf_counter() - system_start

        def format_time(seconds: float) -> str:
            if seconds >= 1.0:
                return f"{seconds:.4f}s"
            elif seconds >= 0.001:
                return f"{seconds * 1000:.3f}ms"
            else:
                return f"{seconds * 1_000_000:.2f}µs"

        print("-----------------------------------------------")
        print(
            f"Total Operations:      {total_operations} Writes, {total_operations} Reads"
        )
        print(f"Main Thread Write Q:   {format_time(queue_time)}")
        print(
            f"Background I/O Wait:   {format_time(time.perf_counter() - commit_start)}"
        )
        print(f"Total System Time:     {format_time(total_time)}")
        print("-----------------------------------------------")
        print(
            f"Mixed Throughput:      {(total_operations * 2) / total_time:.0f} ops/sec"
        )


if __name__ == "__main__":
    benchmark_complex_workload_and_concurrency()


# Test results.

# python stress_complex_json_tests.py
# Generating complex payloads...
# Starting Concurrent Read/Write Stress Test...
# -----------------------------------------------
# Total Operations:      50000 Writes, 50000 Reads
# Main Thread Write Q:   75.4220s
# Background I/O Wait:   40.8361s
# Total System Time:     124.4312s
# -----------------------------------------------
# Mixed Throughput:      804 ops/sec
